### PR TITLE
RavenDB-19061 - Sharding - Stats - Handlers

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19061/Sharding-Stats-Handlers

### Type of change

- Task

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsBytes.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                    [nameof(RequestHandler.Database.Metrics.Attachments.BytesPutsPerSec)] = RequestHandler.Database.Metrics.Attachments.BytesPutsPerSec.CreateMeterData(true, empty)
+                },
+                [nameof(RequestHandler.Database.Metrics.Counters)] = new DynamicJsonValue
+                {
+                    [nameof(RequestHandler.Database.Metrics.Counters.BytesPutsPerSec)] = RequestHandler.Database.Metrics.Counters.BytesPutsPerSec.CreateMeterData(true, empty)
+                },
+                [nameof(RequestHandler.Database.Metrics.TimeSeries)] = new DynamicJsonValue
+                {
+                    [nameof(RequestHandler.Database.Metrics.TimeSeries.BytesPutsPerSec)] = RequestHandler.Database.Metrics.TimeSeries.BytesPutsPerSec.CreateMeterData(true, empty)
+                }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsBytes.cs]: let's use the `ToJson` of `MetricCounters` ?

File: src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsPuts.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                    {
+                        [nameof(RequestHandler.Database.Metrics.Counters.PutsPerSec)] =
+                            RequestHandler.Database.Metrics.Counters.PutsPerSec.CreateMeterData(true, empty)
+                    },
+                [nameof(RequestHandler.Database.Metrics.TimeSeries)] = new DynamicJsonValue
+                {
+                    [nameof(RequestHandler.Database.Metrics.TimeSeries.PutsPerSec)] =
+                        RequestHandler.Database.Metrics.TimeSeries.PutsPerSec.CreateMeterData(true, empty)
+                }
+            });
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsPuts.cs]: same

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetrics.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            public DynamicJsonValue Combine(Memory<BlittableJsonReaderObject> results)
+            {
+                var combined = new DynamicJsonValue();
+
+                for (var i = 0; i < results.Span.Length; i++)
+                {
+                    var result = results.Span[i];
+                    var databaseName = $"{_handler.DatabaseName}${i}";
+                    combined[databaseName] = result.Clone(_context);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetrics.cs]: We don't want to combine here, just proxy the result for (shard, tag)

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsBytes.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        public ShardedStatsHandlerProcessorForGetMetricsBytes([NotNull] ShardedDatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask<DynamicJsonValue> GetDatabaseMetricsAsync(JsonOperationContext context)
+        {
+            using (var token = RequestHandler.CreateOperationToken())
+            {
+                var metrics = await RequestHandler.ShardExecutor.ExecuteParallelForAllAsync(new ShardedStatsHandlerProcessorForGetMetrics.GetShardedDatabaseMetricsOperation(RequestHandler, context, puts: null, bytes: true), token.Token);
+                return metrics;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsBytes.cs]: need to proxy here

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsPuts.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public ShardedStatsHandlerProcessorForGetMetricsPuts([NotNull] ShardedDatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask<DynamicJsonValue> GetDatabaseMetricsAsync(JsonOperationContext context)
+        {
+            using (var token = RequestHandler.CreateOperationToken())
+            {
+                var metrics = await RequestHandler.ShardExecutor.ExecuteParallelForAllAsync(new ShardedStatsHandlerProcessorForGetMetrics.GetShardedDatabaseMetricsOperation(RequestHandler, context, puts: true, bytes: null), token.Token);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsPuts.cs]: same

File: src/Raven.Server/Utils/Metrics/Commands/GetDatabaseMetricsCommand.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
+﻿using System.Net.Http;
+using System.Text;
+using Raven.Client.Http;
+using Sparrow.Json;
+
+namespace Raven.Server.Utils.Metrics.Commands
+{
+    internal class GetDatabaseMetricsCommand : RavenCommand<BlittableJsonReaderObject>
```
Comment (karmeli87) [inline src/Raven.Server/Utils/Metrics/Commands/GetDatabaseMetricsCommand.cs]: Avoid the `BlittableJsonReaderObject` here?

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsBytes.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsBytes.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsPuts.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Handlers/Processors/Stats/StatsHandlerProcessorForGetMetricsPuts.cs]: done

File: src/Raven.Server/Utils/Metrics/Commands/GetDatabaseMetricsCommand.cs (hunk 6 above)
Comment (shiranshalom) [reply src/Raven.Server/Utils/Metrics/Commands/GetDatabaseMetricsCommand.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsPuts.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsPuts.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsBytes.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetricsBytes.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetrics.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Stats/ShardedStatsHandlerProcessorForGetMetrics.cs]: done

**Review by karmeli87 — APPROVED** [review]

**Pull request merged**